### ISSUE TITLE
Implement capability contracts for components

### ISSUE THREAD

**Issue by cmoulliard** [halkyonio/operator#146]
## Description of the problem

The biggest friction we have at the moment with halkyon is that we still have to figure out the wiring manually.

i.e. if a client service needs a backend endpoint, you need to know that the client expects to be injected an endpoint URL in an env variable and know the env name. you will also need to know the URL of the backend service to be able to inject it in the client service as an env link type

The issues I see with this are multiple: 

1. we offer 2 types of links, why should one select one over the other? how do you make that choice? Unclear
2. you need to know what the client app is requiring and in what form (e.g. environment variable named FOO) but you don't currently have access to that information at the halkyon level
3. you need to know the url of the backend service even though that's highly dependent on how that service is deployed and then enter it manually

## Proposition

Based on that, I started thinking that components should have contracts
this would also tie up with the capability concept in a more natural way

a component could define a set of required capabilities and a set of provided capabilities: 
- required capabilities would indicate the type of capability and how the component expect them
- provided capabilities would indicate the kind of services the component is capable of providing
a component couldn't move to the running state without all its required capabilities being satisfied
we could also have optional capabilities but let's not worry about that now

So the client-server interaction would be solved the following way

client specifies: requires a REST capability (i.e. I need an endpoint to connect to) and expect the URL to be provided as ENDPOINT variable
this would be part of the component definition (syntax/semantics TBD)
this could maybe be helped with by dekorate (not sure yet)
this way, the designer of the component / app, lets the rest of the world know what the component needs to be running

On the other side, we have the backend component
there, dekorate can help quite a bit
Based on the analysis we already do on the app, we know that the app exposes endpoints, we could know on what URLs (relative to the app's context) and on what port
So we could automatically generate provided capabilities for the backend component

**Now, how do things tie together?**
well, when you create your app, you will want to link the client with the backend
so hal could look at provided capabilities that match the from component required capabilities and offer to create a "connection" between them
the link could then be automatically created without the user having to know about which env variable to use or how the backend URL is composed
the only thing the user will need to decide is which of the exposed URLs to use
but the rest could (and should) be quite transparent

So is this just some crazy, sleep deprived idea or could this actually make sense?

Example: for the database use case, we could generate a datasource based on the provided contract of the target capability that would be a little more complex to handle
but I could see it working though it might require smarter runtime images (i.e. we might need to be able to run code on the runtime images to do the datasource creation manipulation in a runtime-agnostic way from halkyon)

So a client component could look like:

```
spec:
    deploymentMode: "dev"
    runtime: "spring-boot"
    version: "2.1.6.RELEASE"
    capabilities:
      required:
        - type: "rest"
          parameters:
            - name: "expectedVar"
              value: "ENDPOINT"
      provided:
        - type: "url"
          parameters:
            - name: "context"
              value: "/foo"
            - name: "port"
              value: 8080
```

and the backend component:
```
spec:
    deploymentMode: "dev"
    runtime: "spring-boot"
    version: "2.1.6.RELEASE"
    capabilities:
      required:
        - type: "database"
          parameters:
            - name: "datasourceName"
              value: "myDB"
      provided:
        - type: "rest"
          parameters:
            - name: "context"
              value: "/api"
            - name: "port"
              value: 8080
            - name: "endpoints"
              value:
                - "/foo"
                - "/bar"
```

**Comment by metacosm** [comment]
Moved the issue to the operator project instead…

**Comment by metacosm** [comment]
> I will review the ticket completely but If I follow the idea correctly, then the idea is to let a component to express if it needs something `=required` or if it will propose something `=provided` - correct ?

Yes.

> How then the link will be able to work using such `.spec.capabilities` field ? Example ? @metacosm

To create a link, a client will need to look for capabilities providing what the client component requires. Using `hal` as an example, `hal link` would first ask you to select the component requiring a capability. It would then list all non-satisfied required capabilities (and probably also the possibility at some point to change already satisfied ones). Once selected, `hal` would use the `required` capability metadata to find matches among already known capabilities (i.e. anything with a `provided` field that matches what the component `required` looks for) and, probably in a second stage, offer the possibility to create a new matching capability, think provisioning a DB from the asking component. The user would then inspect / select among all the matches which one to link to. `hal` would then create a link between both component / capability using the client component's `required` metadata to inject the matching information in the component's deployment resource. I suspect that all of this could be done using only secrets as opposed to specifying env variables directly, i.e. a providing component would generate a capability with a secret containing all the metadata needed to consume the capability, adapted to the required metadata of the consuming component… 😅

**Comment by metacosm** [comment]
> I have 2 questions:
> 
>     1. How would the endpoints provided by a component be generated? Would all the endpoints be listed and if not how would one control which ones are included?

Though we could probably list all endpoints using dekorate processing, it probably makes sense to generate only the entry point to the api e.g. context relative to app and port…

>     2. What happens if two different components expose the same endpoint? How would it be decided which link would be followed?

Provided capabilities are component-scoped so no risk of confusion. Also, there is no automatic linking going on, only exposed metadata. It's up to the user to select which endpoint they want to use in their app. We only provide (some) formalization on top of (right now informal) component requirements to help tools discover and help user make their choice.

**Comment by geoand** [comment]
I started playing with the `hal` and `halkyon` today and although I find the concept of the Link between a `Component` and `Capability` awesome, I fail to see the reasoning for a developer would use it between Component.

When I want component `A` to be able to "talk" to Component `B`, I would typically only need a URL and to keep things configurable I would typically just use an env var to set it.
In some scenaria, I might not even need to do that (see Spring profiles for example where a `kubernetes` profile could contain configuration that just points to the DNS of the "downstream" component).

In its current incarnation, the Link between Components is just overhead - since it doesn't provide anything additional to an env var while introducing another layer of abstraction.

**Comment by geoand** [comment]
IIUC, the `linking` part helps an architect have an overall view of the system and that is why we want it? 
I am asking because from a typical developer perspective I can't really see what benefit it brings to the table.

**Comment by metacosm** [comment]
> When I want component `A` to be able to "talk" to Component `B`, I would typically only need a URL and to keep things configurable I would typically just use an env var to set it.

And how would you do that if you don't know what a component is build from, i.e. if you only know about the component, link and capability abstractions? Even if you know about the underlying resources, how would you add the env variable? It certainly is feasible but is creating a link really that far fetched (as opposed to the alternative(s))?

**Comment by metacosm** [comment]
> IIUC, the `linking` part helps an architect have an overall view of the system and that is why we want it?
> I am asking because from a typical developer perspective I can't really see what benefit it brings to the table.

As a developer don't you want to be able to see the dependencies between your services in a more explicit way? As a developer I find it useful to know that my component depends on another one, a relation that the link makes explicit, instead of having to remember that the dependency is hidden in an environment variable with a magic URL in it. 

However, I do agree that the current link implementation is sub-par in that regard. :)

**Comment by geoand** [comment]
It's not far fetched, it's just that it doesn't add any value.
Even if I create a link, I still have to point it to the appropriate coordinates.

For me, just having the ability to add arbitrary env vars without having to think of an extra abstraction layer would be ideal

**Comment by geoand** [comment]
Of course I would LOVE to be able to see the dependencies of my component. 
But I don't want to have to add another abstraction layer which feels unnatural in order to do that.
If there some magical way where I could just say link component A to B without me having to do anything (in the standard case, but allow me to override if needed), then that is something I could get behind.

**Comment by metacosm** [comment]
> For me, just having the ability to add arbitrary env vars without having to think of an extra abstraction layer would be ideal

Again, how do you add env vars if the only abstractions you know are components, links or capabilities? If you know enough about k8s to be able to add env vars directly to your deployment yourself, would you care about halkyon to start with?

**Comment by geoand** [comment]
> > For me, just having the ability to add arbitrary env vars without having to think of an extra abstraction layer would be ideal
> 
> Again, how do you add env vars if the only abstractions you know are components, links or capabilities? If you know enough about k8s to be able to add env vars directly to your deployment yourself, would you care about halkyon to start with?

Of course I would since it provides a ton of more things.

Having to set an env var to change my configuration is something most devs are familiar with regardless of environment. 
I don't see why deploying in Kubernetes makes a difference. If anything having the ability in `hal` to set my environment variable frees me from having to figure out where in massive Kubernetes Deployment YAML file I need to add it for it to actually work :)

**Comment by metacosm** [comment]
> If anything having the ability in `hal` to set my environment variable frees me from having to figure out where in massive Kubernetes Deployment YAML file I need to add it for it to actually work :)

And that's exactly what `hal link create` does, so I'm not sure what is the issue with `link` to add an env var? 
Then again, the point of this discussion is to make links more useful so if you think something doesn't feel right, we need to improve things;

**Comment by geoand** [comment]
I understand that to you `hal link` between components seems obvious. 
What I am saying is that using it instead of the well known and universaly understood concept of the env var without any additional benefit, is just creating confusion in my mind.

How am I supposed to know that when I want to simply set an env var (for pointing a configuration property - a URL in most cases to some other component) I need to use `hal link`?

**Comment by metacosm** [comment]
Because the env var is simply an implementation detail of a relation / link between your component and something else? And because what you really want to do is create a link between your component and something else and NOT just create an env var? To me, the link abstraction feels appropriate in that context because I couldn't care less how the dependency is expressed as long as one, it's explicit and two, it's easy to create it (which is the point I want to improve, i.e. a user shouldn't have to even know about env var or secret).

Why would it be any different to link to capabilities? Why do you think it's more obvious to create a link to a capability? Not trying to argue, just trying to understand…

**Comment by geoand** [comment]
It maybe just be that the current implementation is confusing me - I can't really know for sure however.

In the case of the capability you know exactly what you need to "link" to and fully expect you would need to set multiple properties to get setup. Also, the property names for these capabilites can be somewhat standardised (or clearly documented) so the tool can provide meaningful help.

In the case of linking one component to another, I would typically only provide a URL of downstream components. Often enough, I wouldn't even have to do that since in k8s I can just use the DNS name of the service.
For the tool to provide meaningful help in this case I assume you would need to know the name of the downstream component you are linking to. But if that is the case, then as I said the user can just use the DNS name of the service.

It's possible that this concept is great and I just can't it get it through my thick skull. FWIW, I never really liked or understood the meaning of the similar concept in `odo` either.

**Comment by metacosm** [comment]
Hi @arthurdm, I will look at your design proposal since there might be parallels indeed.

**Comment by cmoulliard** [comment]
> we're trying to solve a very similar problem.

That should be great to work together around this problematic. BTW, another puzzle piece that we should also have a look to simplify the binding is perhaps part of this project - https://github.com/redhat-developer/service-binding-operator

**Comment by cmoulliard** [comment]
> I am interested in your thoughts, and if this approach is something that we could have in common.

I definitively agree to have something in common

**Comment by metacosm** [comment]
Do you think supporting auto-binding à la Spring Boot `@Autowired` would be a good idea? E.g. if a component requires a capability and there is one and only one instance of such capability available on the cluster, should Halkyon automatically link both? Should that be the default behavior if it is implemented or something that would need to be activated explicitly?

**Comment by geoand** [comment]
> Do you think supporting auto-binding à la Spring Boot `@Autowired` would be a good idea? E.g. if a component requires a capability and there is one and only one instance of such capability available on the cluster, should Halkyon automatically link both? Should that be the default behavior if it is implemented or something that would need to be activated explicitly?

I personally think this is an excellent idea and IMHO it should be the default behavior

**Comment by geoand** [comment]
> > For me, just having the ability to add arbitrary env vars without having to think of an extra abstraction layer would be ideal
> 
> Again, how do you add env vars if the only abstractions you know are components, links or capabilities? 

I understand that you want to push the use of these abstractions instead of the Kubernetes ones and that makes sense for what Halkyon is trying to do IMHO. However ENV VARS are way to crucial (because they are used universally across all stacks in order to configure applications) so I think they need to fit in here somehow to allow customization.

> 
> Because the env var is simply an implementation detail of a relation / link between your component and something else

It might an implementation detail from the Halkyon perspective, but from an application developer perspective it's not. An env var is one of the primary means of making an application configurable. So I guess what I am saying is that if I am to define a Link from one component to another, I really really want to have a predictable (and probably customizable) naming scheme so as a developer I can consume the configuration.

Does that make sense?

**Comment by iocanel** [comment]
This is getting a little bit confusing as we are mixing two different topics:

- service discovery
- requirements and capabilities concept.

So, I'll share my views on each topic independently.

## service discovery 

I don't think we should be doing anything at all with service discovery. Kubernetes provides `Services` and the DNS support and the concept of injecting service coordinates via environment variables is too 2014 for me.

The only case where that might make sense is cross-cluster discovery, but still, IMHO, it doesn't justify us putting effort on service discovery.

## requirements and capabilities

The concept of using requirements and capabilities is not something new. People that have spent time with OSGi are very familiar with the concept.

In the OSGi land, the idea was that a component wouldn't start unless or the requirements were satisfied. In the same spirit, a component would get stopped whenever a requirement was no longer satisfied.

This provided two major benefits:

- It was easy to visualize the dependencies of your components.
- You could easily replace a component without leaving the system in an unpredicted state. Everything was gracefully stopped and restarted when the requirements were satisfied again.

In the kubernetes land, we have slightly different concepts.

The closest thing to a capability is a `Service`.
But we have absolutely nothing for a requirement. The requirements are handled internally in the application and in the best-case scenario, a missing requirement will be reflected in the state of the probe.

I can see some value in having resources expressing requirements on a `Service`  if we actually need to advertise it or do something about it (e.g. bringing up and down components). But, for things like [service discovery](#service-discovery) I think that `Service` has all the answers:

- where to connect
- what to do if multiple providers are found etc.

To summarize, I feel that we are trying to solve  problem that is long solved.
The requirements/capabilites concept has value, let's focus in that.

**Comment by metacosm** [comment]
You can see examples of contracts at: https://github.com/halkyonio/operator/blob/capabilities-ng/demo/fruit-backend-sb/halkyon.yml#L34-L51 where the backend component both defines it requires a PostgreSQL DB and provides a REST endpoint which the client component consumes: https://github.com/halkyonio/operator/blob/capabilities-ng/demo/fruit-client-sb/halkyon.yml#L23-L32. 

You'll notice that the client component specifies a "weird" looking env variable: https://github.com/halkyonio/operator/blob/capabilities-ng/demo/fruit-client-sb/halkyon.yml#L31-L32, this is used for the target capability to override the default name of the env var under which the backend REST endpoint URL is published. This allows the application underlying the component to avoid having to adapt to the Halkyon capability mechanism (in essense, that env variable says: "publish your endpoint url into an env var named as I'm telling you, in this case `KUBERNETES_ENDPOINT_FRUIT`").
Though this is not shown in the examples above, where explicit bindings are done using `boundTo` attributes (https://github.com/halkyonio/operator/blob/capabilities-ng/demo/fruit-client-sb/halkyon.yml#L26 and https://github.com/halkyonio/operator/blob/capabilities-ng/demo/fruit-backend-sb/halkyon.yml#L37), we also support auto-binding to a capability if only one capability matching the component's requirement exists on the cluster. While we haven't implemented it yet (and might not do so), it's also conceivable to add an auto-provision attribute for these capabilities that might support it (e.g. databases) where the target capability could be created on demand (e.g. a database would be provisioned) based on the component's requirements if no such capability already exists on the cluster.

**Comment by cmoulliard** [comment]
> You can see examples of contracts at: https://github.com/halkyonio/operator/blob/capabilities-ng/demo/fruit-backend-sb/halkyon.yml#L34-L51 where the backend component both defines it requires a PostgreSQL DB and provides a REST endpoint which the client component consumes: https://github.com/halkyonio/operator/blob/capabilities-ng/demo/fruit-client-sb/halkyon.yml#L23-L32.

Excellent !!

**Comment by cmoulliard** [comment]
> we also support auto-binding to a capability if only one capability matching the component's requirement exists on the cluster.

This mechanism should be documented/described as it represents a great feature such as Spring cloud offers too with their `cf` tool

**Comment by cmoulliard** [comment]
> to add an auto-provision attribute for these capabilities

concerning this point, I think that we should create a separate ticket explaining the strategy that we could support without too much effort
Example: If OLM is deployed and If the `capability` matches an existing operator (= that we could resolve using CSV labels, annotations), then we could issue a subscription.

**Comment by metacosm** [comment]
> > to add an auto-provision attribute for these capabilities
> 
> concerning this point, I think that we should create a separate ticket explaining the strategy that we could support without too much effort
> Example: If OLM is deployed and If the `capability` matches an existing operator (= that we could resolve using CSV labels, annotations), then we could issue a subscription.

That's a slightly different idea, actually. What I meant is the following use case: one component needs a DB but no capability exists for a DB. However, the kubedb capability is available (though no DB has actually been created) and so we could have halkyon create a DB instance based on the component's requirement automatically. This would go one step further above auto-linking: if no DB exists and a capability is installed that can provision instances, then create one. With the current implementation, it wouldn't require much more than an `autoCreate: true` field in the component's associated required capability (and some code at the operator level, of course) to get this working…

**Comment by cmoulliard** [comment]
> That's a slightly different idea, actually. What I meant is the following use case: one component needs a DB but no capability exists for a DB. However, the kubedb capability is available (though no DB has actually been created) and so we could have halkyon create a DB instance based on the component's requirement automatically.

I dont see why what you describe here is different from the OLM/subscription proposition described at #220 as by example a missing prometheus capability could also result into the creation of a subscription CR to tell to OLM to install it and next we could base on default parameters create the `ServiceMonitor` CRD needed to configure prometheus

**Comment by metacosm** [comment]
> > That's a slightly different idea, actually. What I meant is the following use case: one component needs a DB but no capability exists for a DB. However, the kubedb capability is available (though no DB has actually been created) and so we could have halkyon create a DB instance based on the component's requirement automatically.
> 
> I dont see why what you describe here is different from the OLM/subscription proposition described at #220 as by example a missing prometheus capability could also result into the creation of a subscription CR to tell to OLM to install it and next we could base on default parameters create the `ServiceMonitor` CRD needed to configure prometheus

That's completely different :)
In the use case I describe, kubedb is already installed and working, it's just that there is no database instance. There is no need to install any operator, just create a DB instance without the user having the create a capability instance manually. 
What you describe is a completely different use case where you need to install an operator.

**Comment by cmoulliard** [comment]
> just create a DB instance without the user having the create a capability instance manually.

So, what you describe here corresponds also to create a CR for a particular Operator in order to install the service/capability needed.

| Capability | System processing it | CR created | System Operating it |
| --| -- | -- | -- |
| Database/PostgreSQL | Plugin KubeDB | PostgreSQL CR | Kube Controller |
| Metrics/Prometheus | PluginPrometheus | ServiceMonitor | Prometheus Operator |
